### PR TITLE
[ISSUE #5553] Parallelize generating ConsumeQueue

### LINKED ISSUE #5553: [RIP-52] [Task 1] Parallelize generating ConsumeQueue
_No description provided._

### PR DESCRIPTION
**Pull request by nowinkeyy**
**Make sure set the target branch to `develop`**

## What is the purpose of the change

close #5553 

## Brief changelog

XX

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

**Comment by djl394922860** [comment]
Is here have any stress test reports?

**Comment by nowinkeyy** [comment]
> Is here have any stress test reports?

I only ran some simple tests on my computer. It is now known that the previous blocking phenomenon can be avoided, but it is not certain that tps will improve.

**Comment by hzh0425** [comment]
Good job! And please also provide a performance test report to prove this pr! @nowinkeyy

**Comment by lizhanhui** [comment]
> > Is here have any stress test reports?
> 
> I only ran some simple tests on my computer. It is now known that the previous blocking phenomenon can be avoided, but it is not certain that tps will improve.

It would be helpful and persuasive to have A/B test metrics, showing it indeed mitigates the hotspot.

**Comment by nowinkeyy** [comment]
I'll show you the test as soon as I can. : )

**Comment by nowinkeyy** [comment]
I have done some simple tests for your reference only, your comments are welcome. : )

image

**Comment by ShadowySpirits** [comment]
Nice job! Looks like a great improvement. And I have some suggestions for your test.

> 当并发线程较小时生成消息速度有所提升

It sounds weird, if we use a single thread for lock-free and out-of-order dispatch, will we get better performance?

> 原设计处于较后位置的消息会有一定延迟

Can you record the dispatch latency (avg/pt90/pt99) to further illustrate the advantages of your solution?

And dispatch is the core logic of RocketMQ, so it is necessary to use long-running tests to verify its stability, latency, and CPU usage.

**Comment by shy-share** [comment]
@nowinkeyy I suggest that you put up the test code for everyone to see your test method, first of all, to ensure that the method is correct, and secondly, to see whether the results are correct and whether the statistics are complete

**Comment by nowinkeyy** [comment]
> @nowinkeyy I suggest that you put up the test code for everyone to see your test method, first of all, to ensure that the method is correct, and secondly, to see whether the results are correct and whether the statistics are complete

Thanks. I will present more detailed test reports and test code as soon as possible.

**Comment by nowinkeyy** [comment]
This is the latest test data.
image

File: store/src/main/java/org/apache/rocketmq/store/logfile/DefaultMappedFile.java [hunk 1]
```diff
@@ -314,6 +314,42 @@ public boolean appendMessage(final byte[] data, final int offset, final int leng
...
+                buf.position(expectOffset);
+                buf.put(data, 0, data.length);
+            } catch (Throwable e) {
+                log.error("Error occurred when append message to mappedFile.", e);
+            }
+            int length = 0;
+            int currentPos = WROTE_POSITION_UPDATER.get(this);
+            if (currentPos == expectOffset) {
+                // todo how to traverse two mappedByteBuffer
+                for (int index = currentPos; index <= this.mappedByteBuffer.limit(); index += 20) {
```
Comment (lizhimins) [inline store/src/main/java/org/apache/rocketmq/store/logfile/DefaultMappedFile.java:338]: 20 is magic number. Does it support batch cq?

**Review by lizhimins — CHANGES_REQUESTED** [review]

File: store/src/main/java/org/apache/rocketmq/store/logfile/DefaultMappedFile.java (hunk 1 above)
Comment (nowinkeyy) [reply store/src/main/java/org/apache/rocketmq/store/logfile/DefaultMappedFile.java:338]: > 20 is magic number. Does it support batch cq?

It doesn't support bcq. Only ordinary cq are supported for now.

**Pull request closed without merging**